Add queue middleware

adrianlungu opened the pull request:
Hoping this middleware might be useful for others, I'm submitting this PR that adds a Queue Middleware.

Basically, it creates a queue of size `x` and a worker pool of size `y` so no more than `y` requests can be processed at any given time while `x` requests can stack up in the queue.

aldas commented:
just some quick remarks:

* It would be better if you return errors instead of using `c.Error()`. It would allow middlewares up in chain to react on errors
* what happens when next middleware/handler panics - will semaphores be released eventually?

adrianlungu commented:
Good points!

I've adjusted the code, now errors should properly return upwards the chain, and semaphores should be released through the `defer` mechanism. Also added a test case for panics to test that everything's working as expected.

pafuent commented on middleware/queue.go:
```diff
@@ -0,0 +1,115 @@
+			ctxQueue, _ := context.WithTimeout(context.Background(), config.QueueTimeout)
+
+			if err := queueSemaphore.Acquire(ctxQueue, 1); err != nil {
+				return ErrQueueFull
```
if you follow my advice of using the request context as parent of ctxQueue, you should check here if err is DeadlineExceeded, because it could be Canceled if the parent context was canceled.

pafuent commented on middleware/queue.go:
```diff
@@ -0,0 +1,115 @@
+				return ErrQueueFull
+			}
+
+			ctxWorker, _ := context.WithTimeout(ctxQueue, config.WorkerTimeout)
```
Same comment about the CancelFunc

Also, if the parent context is ctxQueue, you will be consuming time of waiting for the workers, with the time spent waiting for the queue (hopes this comment is clear)

pafuent commented on middleware/queue_test.go:
```diff
@@ -0,0 +1,235 @@
+			errQueueTimeout = true
+		}
+
+		if c == http.StatusInternalServerError {
```
This doesn't seems to be useful for testing the middleware code, this error is only returned by the test code (when the procTime is not an int)

pafuent commented on middleware/queue_test.go:
```diff
@@ -0,0 +1,235 @@
+		{95},
+		{95},
+		{95},
+		{120},
```
The 120 and 250 procTime are too near to the Queue and Worker timeouts. Maybe this could lead to random failures when this test are executed. Please run them several times and set values to this procTime that don't generate random failures.

adrianlungu replied:
Unfortunately, I ran them 100 times before committing but I've adjusted the test cases to try to make it as reliable as possible.

pafuent reviewed (commented):
Thanks for the PR!!!

adrianlungu commented on middleware/queue.go:
```diff
@@ -0,0 +1,115 @@
+				config.BeforeFunc(c)
+			}
+
+			ctxQueue, _ := context.WithTimeout(context.Background(), config.QueueTimeout)
```
I initially didn't use the `cancelFunc` as there doesn't seem to be a case where it makes sense to use it, especially if we use the `http.Request.Context()`, since it either gets cancelled by the request, or by the timeout.

But I guess, if it finishes the work, it can be called alongside the semaphores' releases to avoid leaks. 🤔

pafuent reviewed (approved):
LGTM!

adrianlungu commented:
Is there anything else I should look into for this MR to get merged ?

adrianlungu commented:
Any update on this ?
